resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1416,4 +1389,136 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
             }
         }
     }
+
+    private interface DnsResolveChannelProvider {
+        /**
+         * Return the next {@link ChannelFuture} that contains the {@link Channel} that should be used for resolving
+         * a chain of queries.
+         *
+         * @return  the {@link ChannelFuture}
+         */
+        ChannelFuture nextResolveChannel();
+
+        /**
+         * Called once resolve operations completes.
+         *
+         * @param resolveChannelFuture  the {@link ChannelFuture} that was previous obtained
+         *                              via {@link #nextResolveChannel()}.
+         * @param resolveFuture         the {@link Future} that holds the result of the resolve operation.
+         */
+        void resolveComplete(ChannelFuture resolveChannelFuture, Future<?> resolveFuture);
+
+        /**
+         * Close the {@link DnsResolveChannelProvider} and so cleanup resources if needed.
+         */
+        void close();
+    }
+
+    private static final class DnsResolveSameChannelProvider implements DnsResolveChannelProvider {
+
+        private final ChannelFuture resolveChannelFuture;
+
+        DnsResolveSameChannelProvider(Bootstrap bootstrap, SocketAddress localAddress) {
+            SocketAddress address = localAddress == null ? new InetSocketAddress(0) : localAddress;
```

Comment:
Reminder to handle `localAddress` like in #14375. Worth creating a utility for that bcz we will need similar logic in every provider.